AFQ/data.py

```diff
@@ -286,6 +293,1008 @@ def read_templates(resample_to=False):
+        """
+        if self._use_participants_tsv:
+            tsv_key = '/'.join([self.s3_prefix, 'participants.tsv'])
+            s3 = get_s3_client(anon=self.anon)
+
+            def get_subs_from_tsv_key(s3_key):
+                response = s3.get_object(
+                    Bucket=self.bucket,
+                    Key=s3_key
+                )
+
+                return set(pd.read_csv(
+                    response.get('Body')
+                ).participant_id.values)
+
+            subject_set = get_subs_from_tsv_key(s3_key)
+            subjects = list(subject_set)
+        else:
+            s3_prefix = '/'.join([self.bucket, self.s3_prefix])
+            sub_keys = _ls_s3fs(s3_prefix=s3_prefix,
+                                anon=self.anon)['subjects']
+
+            # Just need BIDSLayout for the `parse_file_entities`
+            # method so we can pass dev/null as the argument
+            layout = BIDSLayout(os.devnull, validate=False)
+            subjects = []
+            for key in sub_keys:
+                entities = layout.parse_file_entities(key)
+                subjects.append('sub-' + entities.get('subject'))
+
+        return subjects
+
+    def _download_non_sub_keys(self, directory):
+        fs = s3fs.S3FileSystem(anon=self.anon)
+        for fn in self.non_sub_s3_keys['raw']:
+            fs.get(fn, op.join(directory, op.basename(fn)))
+
+    def download(self, directory, include_non_sub_raw_keys=False,
+                 include_derivs=False, overwrite=False, pbar=True):
+        """Download files for each subject in the study
+
+        Parameters
+        ----------
+        directory : str
+            Directory to which to download subject files
+
+        include_non_sub_raw_keys : bool
+            If True, download all keys in self.non_sub_s3_keys also.
+            This is useful if the non_sub_s3_keys contain files common
+            to all subjects that should be inherited. Default: False
+
+        include_derivs : bool or str
+            If True, download all derivatives files. If False, do not.
+            If a string or sequence of strings is passed, this will
+            only download derivatives that match the string(s) (e.g.
+            ["dmriprep", "afq"]). Default: False
+
+        overwrite : bool
+            If True, overwrite files for each subject. Default: False
+
+        pbar : bool
+            If True, include progress bar. Default: True
+
+        See Also
+        --------
+        AFQ.data.S3BIDSSubject.download
+        """
+        self._local_directories.append(directory)
+        self._local_directories = set(self._local_directories)
+
+        results = [delayed(sub.download)(
+            directory=directory,
+            include_derivs=include_derivs,
+            overwrite=overwrite,
+            pbar=pbar,
+            pbar_idx=idx,
+        ) for idx, sub in enumerate(self.subjects)]
+
+        compute(*results, scheduler='threads')
```

Comment:
Wouldn't you rather use the dask progressbar construct here as well? Instead of the pbar input to `sub.download`?

Reply:
I had trouble getting the nesting to work with the Dask progress bar since we download multiple subjects all at once. So I went with this instead. I admit it's weird to mix the dask progress bar with the tqdm progress bar but I didn't want to spend too much time on that part. But by all means, commits to this PR are welcome.